### ISSUE TITLE
Jamulus 3.9 screen reader focus issue

### ISSUE THREAD

**Issue by BlindRyan** [jamulussoftware/jamulus#2785]
Problem: when arrowing through the list of servers,  it is no longer possible to press right arrow,  and move over in the table to hear other aspects of that server,  such as the delay time,  location,  etc.
 
In version 3.8 (accessible,  when we arrow to a user name,  and then ask Windows for the UIA object with focus,  we obtain the object that contains the server name as the .name field.
Pressing right arrow,  focus moves to the next UIA object,  which contains the response time.  
 
However,  in 3.9, also position on a server name in the list,  what we get when we ask for the UIA object with focus is actually the parent of that server name object,  and seems to relate more to the list in general.  Focus is not reported as being on that particular server.
Pressing right arrow,  focus does not appear to move at all.
 
So,  the root problem is that we are now getting  the wrong object,  and that object does not change as you arrow left and right.

**Comment by ann0see** [comment]
Thanks for your report. I think @chigkim should be able to help out with accessibility related problems.

**Comment by chigkim** [comment]
It looks like QT6 fixed so that VoiceOver on Mac now can at least navigate QTreeWidgetItems in QT6 (not each column though), but maybe it broke in windows...
I'm not sure what we can do until QT fixes QTreeWidgetItems?

As a futile attempt, I tried to set accessible name before the line "// update number of clients text" in connectdlg.cpp.
`lvwServers -> itemWidget ( pCurListViewItem, 1 ) -> setAccessibleName ( pCurListViewItem -> text ( 1 ) );`
However, I get zsh segmentation fault even before GUI opens.
On my accessible build on Mac, I got around it by making each column as a QPushButton, but this of course changes UI.
```
for ( int i = 0; i < lvwServers->columnCount(); i++ )
{
	lvwServers->setItemWidget ( pNewListViewItem, i, new QPushButton ( pNewListViewItem->text ( i ) ) );
}
```

@ann0see, do you have any suggestion?

**Comment by BlindRyan** [comment]
Hi, I am on Windows and wanted to say thanks for looking at it, as someone who is blind it’s great to have a program we can use to jam with others

**Comment by hoffie** [comment]
> It looks like QT6 fixed so that VoiceOver on Mac now can at least navigate QTreeWidgetItems in QT6 (not each column though), but maybe it broke in windows...

The official 3.9.0 build is still on Qt 5.15 though, so this should not be a cause for regression, unless the accessibility Windows build has moved to Qt6 already?

>  On my accessible build on Mac, I got around it by making each column as a QPushButton, but this of course changes UI.

I haven't looked at it, but it sounds like this would not be something which could be integrated by default, but maybe it could be hidden under some general accessibility/screen-reader-friendlyness setting?

I'm still wondering where the change in behavior could be coming from as I don't see any relevant changes between 3.8.2 and 3.9.0 in the connect dialog at all.

@BlindRyan Can you confirm again that 3.8.2 still works as desired? Just to rule out any other, non-Jamulus-related changes on the Windows/screen reader side of things.

@chigkim Does your 3.9.0 accessibility build maybe lack some optimization which 3.8.2 had?

**Comment by BlindRyan** [comment]
Hi,  using Windows 11 with Jamulus  3.8.2 and my screen reader When I am on a server list I am able to use the right arrow to move to each of the items such as the milliseconds, the number of users and the country or region of the server

On my windows ten computer  with jamulus 3.9 I am using the same screen r3eader and when I bring up the exact same list of servers and try to use the right arrow key I get no information

Looking at the jamulus help menu the about qt still says 5 on both 3.8.2 and 3.9

The screen reader I am using if it matters is called Jaws for Windows from www.freedomscientific.com <http://www.freedomscientific.com>

**Comment by chigkim** [comment]
@BlindRyan, just out of curiosity, have you tried with NVDA?

Hmm, maybe then I'm wrong. Since I started building my accessible build 
with QT6, the server list became somewhat accessible on Mac without any 
modification, so I assume it was QT6.

I stopped making a special accessible build for Windows, because all my 
accessibility fixes were merged into the main except for the server list 
on Mac side. That's why I still make accessible build for Mac.

@hoffie, which QT version do 3.8 and 3.9 use? 5.15.x(?)

**Comment by BlindRyan** [comment]
Jamulus 3.8.2 says about qt

This program uses Qt version 5.15.2.

Using NVDA the right arrow key does read out the milliseconds, the number of user as well as the region

 

On Jamulus 3.9 and nvda nothing is spoken when the right arrow is used to move to those items

**Comment by hoffie** [comment]
> I  stopped making a special accessible build for Windows, because all my accessibility fixes were merged into the main

Ah, cool. How did you build your artifacts until 3.8.2? Was it using the same Github infrastructure as in our upstream project? If not, how? Maybe there is another difference.

> which QT version do 3.8 and 3.9 use? 5.15.x(?)

Mac: 3.8: 5.15.2, 3.9: 6.3.1
Win: 3.8: 5.15.2, 3.9: 5.15.2

Doing a Windows Qt6 build should be simple though. We mainly didn't do it as support for "older" Windows versions is somewhat unclear, see #2300.

**Comment by chigkim** [comment]
Hmm, I wonder what changed from 3.8 and 3.9 for Windows though since QT 
version stayed exactly the same.

I've built my accessible build using my local machine and uploaded 
binary as attachment in Github release page.

**Comment by BlindRyan** [comment]
Is that something you want me to download and try to install ? if so can you provide a direct link?

**Comment by BlindRyan** [comment]
This is the message I received from a Jaws Screen reader user who  might explain what is happening

Here is what he wrote to me

Problem: when arrowing through the list of servers,  it is no longer possible to press right arrow,  and move over in the table to hear other aspects of that server,  such as the delay time,  location,  etc.

 

In version 3.8 (accessible,  when we arrow to a user name,  and then ask Windows for the UIA object with focus,  we obtain the object that contains the server name as the .name field.
Pressing right arrow,  focus moves to the next UIA object,  which contains the response time.  

 

However,  in 3.9, also position on a server name in the list,  what we get when we ask for the UIA object with focus is actually the parent of that server name object,  and seems to relate more to the list in general.  Focus is not reported as being on that particular server.
Pressing right arrow,  focus does not appear to move at all.

 

So,  the root problem is that we are now getting  the wrong object,  and that object does not change as you arrow left and right.

**Comment by chigkim** [comment]
I tested 3.9 with NVDA and Jaws.
Here's what you can do.
Server list is a tree view, so when you press right arrow, it expands the selected server, and reveals users on them, and when you press left arrow, it collapses and hide the users.
Things like ping time, number of users, location are not supposed to be keyboard focusable. At least I don't think.
However, in NVDA, once you select a server with up/down, you can use object navigation nvda+shift+right/left to see the number of user, ping time, location.
On Jaws, you can use touch cursor jaws+shift+; and press right/left arrow to review those information.
When you're done, go back to pc cursor by pressing jaws+; twice.
Hope that helps.

**Comment by BlindRyan** [comment]
Hi yep I’ve done that with Jaws but didn’t need to before I updated to 3.9 and then went I went back to 3.8.2 things broke so for now I’ll use the touch cursor in Jaws, I really do appreciate your time looking at this and if I can make a request would it be possible to have a button we can use to toggle the chat functionality? At times the chat can get noisy and disruptive

Thanks again

**Comment by BlindRyan** [comment]
Hi using Jamulus 3.9  on Windows 11 and Jaws 2022 or NVDA when I am on the server llist I am able to use the right arrow to hear the ping and if I right arrow again the number of users and the region

This doesn’t work on Windows ten 

And the above works without having to use the Jaws touch cursor or nvda object navigation on windows 11

**Comment by BlindRyan** [comment]
Hi never min my previous message now the arrow keys won’t work, is it possible to get this functionality to work as it is more efficient than turning on and off the touch cursor, and it did used to work, also any ability to have a option to toggle the chat off during a session

**Comment by chigkim** [comment]
This never worked in main official build. It only worked in my 
accessible build.

For main official build, there isn't really much we can do until QT does 
something about it.

I'm also blind myself using screen readers, so I understand your issue.

I guess I can start releasing special build for Windows again, but for 
official build, using object navigation or touch cursor is best solution 
we have for now.

**Comment by BlindRyan** [comment]
Hopefully they will add some functionality but in the meantime I hired a Jaws scripter to write some scripts that work really well for Jaws in Jamulus

Hee is what the scripts do currently

Once in a server you can do the following

Press control shift p to get your ping and delay

Press control shift u for a list of users, if you press enter on a user you are taken directly to their level slider so you can adjust their levels, much more efficient than tabbing around especially if there are a lot of people 

Control shift o brings up a list of users you can solo

Control shift m brings up a list of the users you can mute

Alt 1 through 0 are for the chat, alt 1 is always the most recent chat

**Comment by hoffie** [comment]
> This never worked in main official build. It only worked in my accessible build. For main official build, there isn't really much we can do until QT does something about it. I'm also blind myself using screen readers, so I understand your issue. I guess I can start releasing special build for Windows again, but for official build, using object navigation or touch cursor is best solution we have for now.

@chigkim Maybe we can add your workaround (I think it changed the UI) to the main build when we put it behind a setting?

**Comment by chigkim** [comment]
@hoffie, we can technically do that, but it's not really an ideal solution either.
It'll break the ability to expand/collapse servers to reveal/hide users with arrows since right/left will navigate to name, ping time, number of users, location instead...
What's even better would be if there's an easy way to make pCurListViewItem as sub class of QWidget.
Then we can just make accessible name to include those infomations with setAccessibleName(...), and it won't change UI either.
I saw QTreeWidget has itemWidget() function that returns QWidget. I was hoping to be able to call setAccessibleName on the returned QWidget value.
However, `lvwServers -> itemWidget ( pCurListViewItem, 1 ) -> setAccessibleName ( pCurListViewItem -> text ( 1 ) );` gives me zsh segmentation fault even before GUI opens.

**Comment by BlindRyan** [comment]
We used to be able to use the right arrow to move all the way over to the region/country of the server and then left arrow back to the server name then press the tab key and get the list of users in that server so could we have one or the other? Either right arrow to expand the users in that server or a press of the tab key to drop us into the list of users?

**Comment by chigkim** [comment]
@BlindRyan, to my knowledge, the official build never behaved the way you described . You're talking about the behavior in my custom accessible build.

**Comment by BlindRyan** [comment]
OK I must’ve been confused I’ve been uninstalling and reinstalling multiple times I also noticed in the three. Nine version there is no sound alert now when somebody enters the server or the room you’re currently in like the accessible build dead I really miss that so I’d like to see that come back as well

**Comment by chigkim** [comment]
It's already in the official build, but the feature is disabled by 
default. You have to check the sound alert in your user profile to 
enable it.

**Comment by pljones** [comment]
Converting to a discussion as the content seems to have wandered.  @chigkim could you please raise an issue making it clear what needs changing.  Thanks.
